containers/src/Data/IntMap/Internal.hs

```diff
@@ -3475,69 +3492,148 @@ instance Read1 IntMap where
-linkWithMask m p1 t1 {-p2-} t2
-  | zero p1 m = Bin p m t1 t2
-  | otherwise = Bin p m t2 t1
+linkWithMask :: Int -> Key -> IntMap a -> Key -> IntMap a -> IntMap a
+linkWithMask m k1 t1 k2 t2
+  | natFromInt k1 < natFromInt k2 = Bin p t1 t2
+  | otherwise = Bin p t2 t1
   where
-    p = mask p1 m
+    p = Prefix (mask k1 m .|. m)
 {-# INLINE linkWithMask #-}
 
 {--------------------------------------------------------------------
   @bin@ assures that we never have empty trees within a tree.
 --------------------------------------------------------------------}
-bin :: Prefix -> Mask -> IntMap a -> IntMap a -> IntMap a
-bin _ _ l Nil = l
-bin _ _ Nil r = r
-bin p m l r   = Bin p m l r
+
+bin :: Prefix -> IntMap a -> IntMap a -> IntMap a
+bin _ l Nil = l
+bin _ Nil r = r
+bin p l r   = Bin p l r
 {-# INLINE bin #-}
 
 -- binCheckLeft only checks that the left subtree is non-empty
-binCheckLeft :: Prefix -> Mask -> IntMap a -> IntMap a -> IntMap a
-binCheckLeft _ _ Nil r = r
-binCheckLeft p m l r   = Bin p m l r
+binCheckLeft :: Prefix -> IntMap a -> IntMap a -> IntMap a
+binCheckLeft _ Nil r = r
+binCheckLeft p l r   = Bin p l r
 {-# INLINE binCheckLeft #-}
 
 -- binCheckRight only checks that the right subtree is non-empty
-binCheckRight :: Prefix -> Mask -> IntMap a -> IntMap a -> IntMap a
-binCheckRight _ _ l Nil = l
-binCheckRight p m l r   = Bin p m l r
+binCheckRight :: Prefix -> IntMap a -> IntMap a -> IntMap a
+binCheckRight _ l Nil = l
+binCheckRight p l r   = Bin p l r
 {-# INLINE binCheckRight #-}
 
 {--------------------------------------------------------------------
-  Endian independent bit twiddling
+  Branching
 --------------------------------------------------------------------}
 
--- | Should this key follow the left subtree of a 'Bin' with switching
--- bit @m@? N.B., the answer is only valid when @match i p m@ is true.
-zero :: Key -> Mask -> Bool
-zero i m
-  = (natFromInt i) .&. (natFromInt m) == 0
-{-# INLINE zero #-}
+-- | A @MapMapBranch@ is returned by 'mapMapBranch' to indicate how two @Bin@s
+-- relate to each other.
+--
+-- Consider that @A@ and @B@ are the @Bin@s whose @Prefix@es are given to
+-- @mapMapBranch@ as the first and second arguments respectively.
+data MapMapBranch
+  = ABL  -- ^ A contains B in the left child
+  | ABR  -- ^ A contains B in the right child
+  | BAL  -- ^ B contains A in the left child
+  | BAR  -- ^ B contains A in the right child
+  | EQL  -- ^ A and B have equal prefixes
+  | NOM  -- ^ A and B have prefixes that do not match
+
+-- | Calculates how two @Bin@s relate to each other by comparing their
+-- @Prefix@es.
+
+-- Notes:
+-- * pw .|. (pw-1) sets every bit below the mask bit to 1. This is the greatest
+--   key the Bin can have.
+-- * pw .&. (pw-1) sets the mask bit and every bit below it to 0. This is the
+--   smallest key the Bin can have.
+--
+-- First, we compare the prefixes to each other. Then we compare a prefix
```

Comment:
What do we learn by comparing the prefixes to each other? That's not exactly an obvious thing to do, and needs documentation. Here's my understanding so far:

`p1` is the smallest permissible value in the right child of its `Bin`. If `pw1 < pw2`, then (according to `left`) `p1` is to the left of the split created by `p2`'s `Bin`.

How exactly does that help?

Reply:
I don't know how to explain this in a way that conveys more information than the code.

> If `pw1 < pw2`, then (according to `left`) `p1` is to the left of the split created by `p2`'s `Bin`.

That's correct.

> How exactly does that help?

It helps because it narrows down the possibilities from 6 to 3. For instance, if `p1`'s `Bin` is in the right child of `p2`'s `Bin` it implies that `pw1 > pw2`, since `pw2` is the smallest permissible value of the right child of `p2`'s `Bin`. Which is ruled out if `pw1 < pw2` is true.